### ISSUE TITLE
Incorrect type narrowing in the type(x) is self.__class__ (regression in 1.1.331)

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#6129]
Before 331 i was happy with `type(other) is self.__class__`, but now i must use isinstance and i think its bad for places like this. 

```python
@dataclass(eq=False)
class C:
    x: int

    def __eq__(self, other: object) -> bool:
        if type(other) is self.__class__:
            return self.x == other.x #  error: Cannot access member "x" for type "object"

        return NotImplemented
```

**Comment by minmax** [comment]
I also discovered strange behavior with type narrowing, types appear out of nowhere.

```python
from requests import PreparedRequest

body = PreparedRequest().body
reveal_type(body)  # Type of "body" is "bytes | str | None"
if isinstance(body, bytes):
    reveal_type(body)  # Type of "body" is "bytes"
    body = body.decode()
    reveal_type(body)  # Type of "body" is "str"

reveal_type(body)  # Type of "body" is "str | bytearray | memoryview | None"
```

**Comment by erictraut** [comment]
Thanks for reporting the issue. This is a duplicate of #6118. It will be fixed in the next release.

For the second code sample, this is intended behavior. The type annotation `bytes` has historically implied `bytes | bytearray | memoryview` (similar to how the type annotation `float` implies `float | int`). That means a value annotated as `bytes` might actually be a `bytearray` or `memoryview` at runtime. If you use an `isinstance` type guard to validate that it is actually a `bytes` value, it may still be a `bytesarray | memoryview` in the fall-through (else) case. That's where these types arise (seemingly out of nowhere).

[PEP 688](https://peps.python.org/pep-0688/#no-special-meaning-for-bytes) recommends that type checkers no longer treat `bytes` in this manner, but some existing code bases still rely on this behavior. I recently added a configuration switch called `disableBytesTypePromotions`. If you set this to true in your config file, pyright will _no longer_ assume that a `bytes` annotation should also imply `bytearray | memoryview`. The `disableBytesTypePromotions` setting is true by default in "strict" type checking mode, but it's false by default in normal type checking mode. At some point in the future, I hope to default this to true all the time.

**Comment by erictraut** [comment]
This is included in pyright 1.1.332, which I just published. It will also be included in a future release of pylance.
